--- a/kernel/cgroup/cgroup.c	2019-07-11 14:27:16.000000000 +0200
+++ b/kernel/cgroup/cgroup.c	2019-07-30 02:33:32.000000000 +0200
@@ -1790,7 +1790,7 @@ static struct dentry *cgroup_mount(struc
 
 	if (fs_type == &cgroup2_fs_type) {
 		if (data && *(char *)data != '\0') {
-			pr_err("cgroup2: unknown option \"%s\"\n", (char *)data);
+			/* pr_err("cgroup2: unknown option \"%s\"\n", (char *)data); */
 			put_cgroup_ns(ns);
 			return ERR_PTR(-EINVAL);
 		}
